asapdiscovery-ml/asapdiscovery/ml/inference/inference.py

```diff
@@ -0,0 +1,108 @@
+import logging
+from typing import Optional
+
+import torch
+from asapdiscovery.ml.utils import build_model, load_weights
+
+# static import of models from base yaml here
+from asapdiscovery.ml.weights import (
+    all_models_spec,
+    fetch_weights,
+    fetch_weights_from_spec,
+)
+
+
+class InferenceBase:
+    """
+    Inference base class for PyTorch models in asapdiscovery.
+
+    Parameters
+    ----------
+    model_name : str
+        Name of model to use.
+    model_type : str
+        Type of model to use.
+    from_spec : bool, default=True
+        Whether to fetch weights from asapdiscovery.ml spec file or from local file.
+    build_model_kwargs : Optional[Dict], default=None
+        Keyword arguments to pass to build_model function.
+    device : str, default='cpu'
+        Device to use for inference.
+
+
+    Methods
+    -------
+    predict
+        Predict on a batch of data.
+
+    """
+
+    def __init__(
+        self,
+        model_name: str,
+        model_type: str,
+        from_spec: bool = True,
+        build_model_kwargs: Optional[Dict] = None,
+        device: str = "cpu",
+    ):
+        logging.info(f"initializing {self.__class__.__name__} class")
+
+        self.device = device
+        logging.info(f"using device {self.device}")
+
+        self.model_name = model_name
+        self.model_type = model_type
+        logging.info(f"using model {self.model_name} of type {self.model_type}")
+
+        # load model weights or fetch them
+        if from_spec:
+            logging.info("using weights from asapdiscovery.ml models.yaml spec file")
+            weights, types = fetch_weights_from_spec(all_models_spec, model_name)
+            if types[model_name] != self.model_type:
+                raise ValueError(
+                    f"Model type {types[model_name]} does not match {self.model_type}"
+                )
+            self.weights = weights[model_name]
+        else:
+            logging.info("using weights from specified local file")
+            self.weights = fetch_weights(model_name)
+        logging.info(f"found weights {self.weights}")
+
+        # build model, this needs a bit of cleaning up in the function itself.
+        self.model, _ = build_model(model_type, **build_model_kwargs)
+        logging.info(f"built model {self.model}")
+
+        # load weights
+        self.model = load_weights(self.model, self.weights)
```

Comment:
[ probably out of scope for this PR ] there should be a test in this function to check whether the weights are compatible with the model. In some cases users will load the 'wrong' weights that are not loadable. If pytorch is anything like TF this will throw a fairly cryptic error, so we should probably catch it more elegantly.

Reply:
I have a function `utils.load_weights` that I made to handle loading weights for an mtenn model, maybe we can build that out a bit and use that?

Reply:
we are using `load weights` on L76 happy to build some guardrails into it.

Reply:
ah whoops, completely missed that somehow, my bad!